improve(voter-app): update app to work with newest voting version 1.2.2

daywiss opened the pull request:
Signed-off-by: David Adams <david@umaproject.org>

<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.
-->

<!--
  Title
  Please include a concise title that briefly describes the change.
  Titles should follow https://www.conventionalcommits.org/.
  They should also be in the present simple tense.

  Examples:
  
  feat(dvm): adds a new function to compute voting rewards offchain
  fix(monitor): fixes broken link in liquidation log
  feat(voter-dapp): adds countdown timer component to the header
  build(solc): updates solc version to 0.6.12
  improve(emp-client): parallelizes web3 calls to improve performance

  For examples of other types (feat, fix, build, improve) and what they mean, take a look at the angular list:
  https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type

  See https://github.com/UMAprotocol/protocol/blob/master/CONTRIBUTING.md#conventional-commits for more details on PR
  title expectations.
-->


**Motivation**

#2371


**Summary**

Adds support for latest contract abi with ancillary data and fixes some bugs.

**Details**
- updates package file to latest abis
- uses ancillary data in calls when needed. also sets it to 0x since there is no ability to specify in this pr
- fixes infinite loop in requesting resolved requests
- tested with run_test script for 2key and non 2key scenarios
- "fixes" bug when decoding encrypted data which i ran into again, see #1969. this was resolved by updating the library version of `eth-crypto`

**Issue(s)**

<!-- This PR must fix or refer to one or more issues. Please list them here. -->
Fixes #2371
Fixes #1969

daywiss commented on packages/voter-dapp/src/ResolvedRequests.js:
```diff
@@ -84,8 +84,9 @@ function ResolvedRequests({ votingAccount }) {
     "Governor",
     "NewProposal",
-    useMemo(() => ({ fromBlock: 0 }))
+    useMemo(() => ({ fromBlock: 0 }), [])
```
This is what was causing infinite loop

nicholaspai replied:
Why does this fix the infinite loop? Really elegant solution if so

daywiss replied:
i think because if you dont unclude the `[]` it will run every render, and its probably causing a rerender when it runs. by adding the `[]` it means just run once.

nicholaspai replied:
I think the second arg to any of the `useX` hooks is an array of dependencies, so I guess if you leave out the array of dependency then it will always re-render. If you explicitly set the dependencies to non with `[]` then it should only render once.

daywiss replied:
yes thats my understanding and explained a lot better

daywiss commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -248,11 +248,17 @@ function ActiveRequests({ votingAccount, votingGateway, snapshotContract }) {
+      encryptedVoteMap[
+        // Ancillary data is undefined or null in these "returnValues" so we add a default string of "0x".
+        // If not added, the key will improperly index and the votes will not be found.
+        toPriceRequestKey(ev.returnValues.identifier, ev.returnValues.time, ev.returnValues.ancillaryData || "0x")
```
this is ugly but for some reason ev.returnValues.ancillaryData is undefined, even though it was sent as 0x on the vote. This causes an issue if you dont default it to '0x'

nicholaspai replied:
I experienced this also when testing the `OptimisticOracleClient` and I think using the `var || 0x` is better than how I addressed it which was the ternary operator: `var ? var : 0x`

nicholaspai commented on packages/core/scripts/mainnet/ProposeAdmin.js:
```diff
@@ -32,7 +32,7 @@ async function propose(callback) {
     console.group(`Proposer account: ${signingAccount}`);
     if (signingAccount !== "0x2bAaA41d155ad8a4126184950B31F50A1513cE25") {
-      throw new Error("Cannot propose admin votes from this account");
+      console.warn("Warning: Cannot propose admin votes if using a mainnet provider from the current account");
```
+1

nicholaspai commented on packages/voter-dapp/run_tests.sh:
```diff
@@ -177,7 +178,7 @@ echo '#                                                                  #'
 echo '#                                                                  #'
 echo -e '####################################################################\n'
-echo "- Follow this tutorial to set up your 2Key wallet via the voter dApp: https://docs.umaproject.org/tutorials/voting-2key. Use your current address as the Cold Wallet Address and transfer your tokens to the proxy contract. You will be prompted to connect again to the voter dApp, but if this doesn't happen automatically then you can simply refresh the page. The Resolved Requests will no longer ...
+echo "- Follow this tutorial to set up your 2Key wallet via the voter dApp: https://docs.umaproject.org/uma-tokenholders/voting-2key. Use your current address as the Cold Wallet Address and transfer your tokens to the proxy contract. You will be prompted to connect again to the voter dApp, but if this doesn't happen automatically then you can simply refresh the page. The Resolved Requests will no ...
```
Thanks for updating this script!

nicholaspai commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -73,9 +73,10 @@ const editStateReducer = (state, action) => {
+// toPriceRequestKey( identifier,time, ancillarydata )
+const toPriceRequestKey = (...args) => args.join(",");
+// toVotingAccountAndPriceRequestKey( votingAccount, identifier, time, ancillarydata )
+const toVotingAccountAndPriceRequestKey = (...args) => args.join(",");
```
`toPriceRequestKey` and `toVotingAccountAndPriceRequestKey` now implement the same function. I find it a bit more readable to define the different vars that will be joined with commas, similar to how you define it in the comments. Why'd you change it?

daywiss replied:
Yea good point, i was a bit on the fence about this. The reason why is that it can accept arbitrary arguments now, for instance with or without ancillary data if needed, though this might not be a good enough reason.  I dont mind changing it back.

nicholaspai replied:
> The reason why is that it can accept arbitrary arguments now, for instance with or without ancillary data if needed

This is definitely useful, but I think its confusing to have two functions implement the same logic. So if you want the flexibility of an unspecified amount of args, maybe consolidate the functions into one `toPriceRequestKey`

nicholaspai commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -312,9 +318,13 @@ function ActiveRequests({ votingAccount, votingGateway, snapshotContract }) {
-            );
+            try {
+              return JSON.parse(
+                await decryptMessage(decryptionKeys[account][currentRoundId].privateKey, voteStatus.committedValue)
```
Does this still fail since we last talked? Catching this error might confuse users in case the messages cannot be decrypted, but I suppose its better than crashing the App.

daywiss replied:
This was failing for me until I bumped the eth-crypto version from 1.6.0 to 1.7.0 in the common package. I dont know if this will break other things, but doesnt seem like it so far.  I left the try catch in since this wasnt really being handled.

nicholaspai replied:
Got it. I'm OK with leaving that here so the App doesn't crash, but its still going to be jarring for the user if they can't see their committed value (they will wonder, "did my commit go through correctly"?)

daywiss replied:
yea thats a great point but im not aware of any infrastructure within the voter dapp to alert the user to errors. if you have a suggestion on how to do this LMK

nicholaspai commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -353,7 +363,8 @@ function ActiveRequests({ votingAccount, votingGateway, snapshotContract }) {
           price: decryptedCommits[index].price.toString(),
-          salt: decryptedCommits[index].salt
+          salt: decryptedCommits[index].salt,
+          ancillaryData: "0x"
```
Shouldn't we be reading the ancillary data from `decryptedCommits` or is the decryption not working?

nicholaspai replied:
Actually, see my comment below in the `commits.push` code.

daywiss replied:
you might be right about this, probably a mistake.

nicholaspai commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -387,6 +398,7 @@ function ActiveRequests({ votingAccount, votingGateway, snapshotContract }) {
       commits.push({
         identifier: pendingRequests[index].identifier,
         time: pendingRequests[index].time,
+        ancillaryData: "0x",
```
So clearly the committed `ancillaryData` and the revealed `ancillaryData` have to be identical. Furthermore, in the future we might want to change this field dynamically depending on the type of vote we're having. Therefore I think its sensible to set a constant ANCILLARY_DATA object at the file level that both this and reveals can read from. This way we can more easily upgrade the ancillary data logic to be dynamic based on types of votes.

daywiss replied:
The next PR will be to add a field so the user can change ancillary data, but yea the default should at least be a constant.

nicholaspai commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -408,7 +420,8 @@ function ActiveRequests({ votingAccount, votingGateway, snapshotContract }) {
         pendingRequests[index].identifier,
-        pendingRequests[index].time
+        pendingRequests[index].time,
+        pendingRequests[index].ancillaryData
```
Wait, why do you use `pendingRequests[index].ancillaryData` here but hardcode `0x` above?

daywiss replied:
Yea I think i made some mistakes here, good catch

nicholaspai reviewed (commented):
The changes are well implemented and overall its not much code which is awesome! I have a comment about how you set/get `ancillaryData` for commits and reveals.

Also, if this PR goes in will the voter dApp no longer be compatible with the current DVM? If so, we should discuss how to handle votes until we migrate the DVM. Perhaps we deploy a version of the web app to a new URL.

daywiss commented:
@nicholaspai thanks for the review. I addressed your concerns. I'm still not 100% sure of this but it does work correctly in the test. I think the next addition when adding user input for ancillary data should shake out further issues.

daywiss commented:
> The changes are well implemented and overall its not much code which is awesome! I have a comment about how you set/get `ancillaryData` for commits and reveals.
> 
> Also, if this PR goes in will the voter dApp no longer be compatible with the current DVM? If so, we should discuss how to handle votes until we migrate the DVM. Perhaps we deploy a version of the web app to a new URL.

I think @mrice32 mentioned he has a plan for this release and noone seemed too concerned about this... Maybe matt can comment

mrice32 commented:
> > The changes are well implemented and overall its not much code which is awesome! I have a comment about how you set/get `ancillaryData` for commits and reveals.
> > Also, if this PR goes in will the voter dApp no longer be compatible with the current DVM? If so, we should discuss how to handle votes until we migrate the DVM. Perhaps we deploy a version of the web app to a new URL.
> 
> I think @mrice32 mentioned he has a plan for this release and noone seemed too concerned about this... Maybe matt can comment

@daywiss I think the simplest thing to do is just wait until the upgrade is over to merge this.

pemulis commented:
While we wait on the upgrade to merge this, should we resolve the merge conflicts?

daywiss commented:
> While we wait on the upgrade to merge this, should we resolve the merge conflicts?

Ill fix conflicts when we are ready to go with this pr. shouldnt take long

daywiss commented on packages/core/scripts/local/RequestOraclePrice.js:
```diff
@@ -61,13 +62,24 @@ async function run(deployedFinder, identifier, timeString) {
+    if (ancillaryDataInBytes) {
+      priceExists = await deployedVoting.hasPrice(identifierInBytes, timeInBN, ancillaryDataInBytes);
+    } else {
+      priceExists = await deployedVoting.hasPrice(identifierInBytes, timeInBN);
```
this script asks for new prices. im allowing this script to run either function to ensure when testing voter dapp both situations work ( and it does)

daywiss replied:
depends on what you mean by does it work. I've updated the test script to create price requests that hit both of these branches, and it works as expected: price requests are made with and without ancillary data. Its possible that line 70 is still calling the ancillary version of the function though, which could be giving me similar behavior.

daywiss commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -140,7 +141,7 @@ function ActiveRequests({ votingAccount, votingGateway, snapshotContract }) {
       }
     }
-  }, [allPendingRequests, IDENTIFIER_BLACKLIST, showSpamRequests]);
+  }, [allPendingRequests, hexToUtf8, showSpamRequests]);
```
modified this due to build warnings. most warnings are gone now

nicholaspai replied:
Super strange that this `useEffect()` hook would depend on the value of `hexToUtf8`, which is a function?

daywiss replied:
its not that it doesnt depend on it, but react warns that mutations to `IDENTIFIER_BLACKLIST` will not re-render component, so it isnt needed
![image](https://user-images.githubusercontent.com/4429761/109739315-6cc41500-7b97-11eb-8c48-3b9cea1f2805.png)

daywiss commented on packages/voter-dapp/src/DesignatedVotingDeployment.js:
```diff
@@ -53,7 +53,11 @@ function DesignatedVotingDeployment({ votingAccount }) {
-      <a href="https://docs.umaproject.org/uma/oracle/voting_with_UMA_2-key_contract.html" target="_blank">
+      <a
+        href="https://docs.umaproject.org/uma/oracle/voting_with_UMA_2-key_contract.html"
+        rel="noopener noreferrer"
```
this was added to stop compiler warning

nicholaspai replied:
Why does the `rel` fix the compiler warning OOC?

daywiss replied:
![image](https://user-images.githubusercontent.com/4429761/109701524-1dfb8880-7b61-11eb-98e2-6af29b5cd2aa.png)

daywiss commented on packages/voter-dapp/src/ResolvedRequests.js:
```diff
@@ -151,7 +152,7 @@ function ResolvedRequests({ votingAccount }) {
       }
     }
-  }, [allResolvedEvents, IDENTIFIER_BLACKLIST, showSpamRequests]);
+  }, [hexToUtf8, allResolvedEvents, showSpamRequests]);
```
again recommended change by compiler

nicholaspai replied:
Makes no sense...

daywiss replied:
![image](https://user-images.githubusercontent.com/4429761/109701672-46838280-7b61-11eb-840a-02d669b39d7d.png)

mrice32 replied:
The second part confuses me @daywiss -- can you shed a light on why the blacklist "doesn't re-render the component"?

daywiss replied:
im guessing its because this is state defined outside of a react component, so react cant react to changes to this variable. on top of that its a constant as well, so its kind of moot, it should never change.

nicholaspai commented on packages/voter-dapp/package.json:
```diff
@@ -9,7 +9,6 @@
     "@material-ui/styles": "^4.3.0",
     "@uma/common": "^2.0.1",
     "@uma/core": "^2.0.1",
-    "@uma/core-1-2-0": "npm:@uma/core@1.2.0",
```
What did this binding do before?

daywiss replied:
this was due to @mrice32 PR #2586. allowed us to load in new and old version of contract into app so we had both abis avaialble, or something like that

nicholaspai commented on packages/voter-dapp/run_tests.sh:
```diff
@@ -39,18 +39,14 @@ echo -e '####################################################################\n'
 # This forces the voter dapp to deploy the _new_ voting contract with all the other contracts unchanged, which will
 # force it to test the old 2key contract with the new voting contract.
 yarn truffle compile
-ARTIFACT_PATH_1_2_0=$(node -p "require.resolve('@uma/core-1-2-0/build/contracts/Voting.json')")
```
Great to see this complexity removed.

nicholaspai commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -417,7 +439,8 @@ function ActiveRequests({ votingAccount, votingGateway, snapshotContract }) {
             price,
-            identifierPrecision
+            identifierPrecision,
+            ancillaryData
```
Nice adding this here, not sure how you would have thought to do so without remembering about this button but +1

daywiss replied:
@nicholaspai actually your run_tests script is very thorough and once i added requests with both ancillary and non ancillary data I had to do a lot more updates. pretty much every change in this pr was found thanks to that script 👍

nicholaspai commented on packages/voter-dapp/src/ResolvedRequests.js:
```diff
@@ -128,10 +129,10 @@ function ResolvedRequests({ votingAccount }) {
     setHasSpamRequests(false);
-
-    if (allResolvedEvents) {
+    if (allResolvedEvents && allResolvedEvents.length) {
```
+1

nicholaspai replied:
Can you not do `allResolvedEvents?.length`?

daywiss replied:
interesting, yea i havent used this new syntax yet but its a good candidate for it. can update

nicholaspai commented on packages/voter-dapp/src/RetrieveRewards.js:
```diff
@@ -225,7 +232,7 @@ function RetrieveRewards({ votingAccount }) {
         .catch(err => console.error(`retrieveRewards.call failed:`, err));
     }
-  }, [votingAccount, rewardsTxn.oldestUnclaimedRound, rewardsTxn.toRetrieve]);
+  }, [web3.utils, drizzle.contracts.Voting, votingAccount, rewardsTxn.oldestUnclaimedRound, rewardsTxn.toRetrieve]);
```
Also super strange to see `web3.utils` here

daywiss replied:
yea... i was just following the compiler warnings here, i dont really understand why it wanted this

nicholaspai reviewed (approved):
Really clean code and its great to see a bunch of hack fixes going away!

daywiss commented on packages/common/package.json:
```diff
@@ -42,7 +42,7 @@
     "chalk-pipe": "^3.0.0",
     "dotenv": "^6.2.0",
-    "eth-crypto": "^1.6.0",
+    "eth-crypto": "^1.7.0",
```
not sure if this is strictly necessary, but running locally without this change doesnt seem to be an option

mrice32 commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -248,11 +248,23 @@ function ActiveRequests({ votingAccount, votingGateway, snapshotContract }) {
+        toPriceRequestKey(
+          ev.returnValues.identifier,
+          ev.returnValues.time,
+          ev.returnValues.ancillaryData || DEFAULT_ANCILLARY_DATA
```
Is ev.returnValues.ancillaryData null when the bytes field is empty? I would expect it to be `0x`...

mrice32 commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -312,9 +324,13 @@ function ActiveRequests({ votingAccount, votingGateway, snapshotContract }) {
+                await decryptMessage(decryptionKeys[account][currentRoundId].privateKey, voteStatus.committedValue)
+              );
+            } catch (err) {
+              return "";
```
If we're going to catch this error, can we return a string value that indicates the error? I think this should be distinguishable from no vote in some way. Something like "ERROR: DECRYPTION FAILED"? Would be ideal to show that in the "value" field in the dapp. Another (slightly less ideal) solution could be to print out the error using console.error so we can diagnose the issue in prod. Otherwise the error gets swallowed and there's no way to find the cause.

daywiss replied:
yes, this should log to console.error since theres no way to really notifiy the user in this app

mrice32 commented on packages/voter-dapp/src/ActiveRequests.js:
```diff
@@ -345,13 +361,15 @@ function ActiveRequests({ votingAccount, votingGateway, snapshotContract }) {
       .catch(err => console.log("snapshot error", err));
   };
 
+  // keccak256(abi.encodePacked(price, salt, msg.sender, time, ancillaryData, roundId, identifier))
```
commented code?

mrice32 reviewed (approved):
LGTM!
